File: src/lib/validations/project.ts
import { z } from "zod";
import {
  ProjectType,
  ProjectStatus,
  ProjectPriority,
} from "@prisma/client";
import { optionalTrimmedString } from "@/lib/validations/shared";

export const ProjectFormSchema = z.object({
  name: z.string().trim().min(2, "Project name must be at least 2 characters"),
  clientId: z.string().trim().min(1, "Client is required"),
  projectType: z.nativeEnum(ProjectType),
  status: z.nativeEnum(ProjectStatus),
  priority: z.nativeEnum(ProjectPriority),
  description: optionalTrimmedString,
  intakeSummary: optionalTrimmedString,
  scope: optionalTrimmedString,
  techStack: optionalTrimmedString,
  domainName: optionalTrimmedString,
  hostingInfo: optionalTrimmedString,
  startDate: optionalTrimmedString,
  ownerUserId: optionalTrimmedString,
  tags: z.array(z.string().trim().min(1)).default([]),
});

export type ProjectFormData = z.infer<typeof ProjectFormSchema>;

File: backend/app/schemas/project.py
from pydantic import BaseModel, field_validator
from datetime import date, datetime
from decimal import Decimal
from app.models.project import ProjectType, ProjectStatus, Priority


class ProjectCreate(BaseModel):
    client_id: str
    name: str
    project_type: ProjectType
    status: ProjectStatus = ProjectStatus.LEAD
    priority: Priority = Priority.MEDIUM
    description: str | None = None
    intake_summary: str | None = None
    scope: str | None = None
    tech_stack: str | None = None
    domain_name: str | None = None
    hosting_info: str | None = None
    start_date: date | None = None
    owner_user_id: str | None = None
    tags: list[str] = []
    tools_used: list[str] | None = None
    delivery_form: str | None = None
    recurring_fee: Decimal | None = None

    @field_validator("name")
    @classmethod
    def name_min_length(cls, v: str) -> str:
        if len(v.strip()) < 2:
            raise ValueError("Name must be at least 2 characters")
        return v.strip()


class ProjectUpdate(BaseModel):
    name: str | None = None
    project_type: ProjectType | None = None
    status: ProjectStatus | None = None
    priority: Priority | None = None
    description: str | None = None
    intake_summary: str | None = None
    scope: str | None = None
    tech_stack: str | None = None
    domain_name: str | None = None
    hosting_info: str | None = None
    start_date: date | None = None
    owner_user_id: str | None = None
    tags: list[str] | None = None
    tools_used: list[str] | None = None
    delivery_form: str | None = None
    recurring_fee: Decimal | None = None


class ProjectResponse(BaseModel):
    id: str
    client_id: str
    name: str
    slug: str
    project_type: str
    status: str
    priority: str
    description: str | None
    intake_summary: str | None
    scope: str | None
    tech_stack: str | None
    domain_name: str | None
    hosting_info: str | None
    start_date: date | None
    owner_user_id: str | None
    tags: list[str]
    tools_used: list[str] | None
    delivery_form: str | None
    recurring_fee: Decimal | None
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class ProjectDetailResponse(ProjectResponse):
    client: dict | None = None
    owner: dict | None = None
    communications: list = []
    change_requests: list = []
    repositories: list = []
    links: list = []


class ProjectFilterParams(BaseModel):
    status: ProjectStatus | None = None
    client_id: str | None = None
    priority: Priority | None = None
